framework/vulkan_sample.h

```diff
@@ -99,244 +101,274 @@ namespace vkb
+	typename std::conditional<CppBindings, vkb::rendering::HPPRenderContext, vkb::RenderContext>::type       &get_render_context();
+	typename std::conditional<CppBindings, vkb::rendering::HPPRenderContext, vkb::RenderContext>::type const &get_render_context() const;
+	bool                                                                                                      has_render_context() const;
 
-	/**
-	 * @brief Additional sample initialization
-	 */
-	bool prepare(const ApplicationOptions &options) override;
+	/// <summary>
+	/// PROTECTED VIRTUAL INTERFACE
+	/// </summary>
+  protected:
+	// from Application
+	void input_event(const InputEvent &input_event) override;
+	void finish() override;
+	bool resize(uint32_t width, uint32_t height) override;
 
 	/**
 	 * @brief Create the Vulkan device used by this sample
 	 * @note Can be overridden to implement custom device creation
 	 */
-	virtual void create_device();
+	virtual typename std::conditional<CppBindings, std::unique_ptr<vkb::core::HPPDevice>, std::unique_ptr<vkb::Device>>::type create_device(
+	    typename std::conditional<CppBindings, vkb::core::HPPPhysicalDevice, vkb::PhysicalDevice>::type &gpu);
 
 	/**
 	 * @brief Create the Vulkan instance used by this sample
 	 * @note Can be overridden to implement custom instance creation
 	 */
-	virtual void create_instance();
+	virtual typename std::conditional<CppBindings, std::unique_ptr<vkb::core::HPPInstance>, std::unique_ptr<vkb::Instance>>::type create_instance(
+	    bool headless);
 
 	/**
-	 * @brief Main loop sample events
+	 * @brief Override this to customise the creation of the render_context
 	 */
-	void update(float delta_time) override;
+	virtual void create_render_context();
 
-	bool resize(uint32_t width, uint32_t height) override;
+	/**
+	 * @brief Prepares the render target and draws to it, calling draw_renderpass
+	 * @param command_buffer The command buffer to record the commands to
+	 * @param render_target The render target that is being drawn to
+	 */
+	virtual void draw(vkb::core::HPPCommandBuffer &command_buffer, vkb::rendering::HPPRenderTarget &render_target);
+	virtual void draw(CommandBuffer &command_buffer, RenderTarget &render_target);
 
-	void input_event(const InputEvent &input_event) override;
+	/**
+	 * @brief Samples should override this function to draw their interface
+	 */
+	virtual void draw_gui();
 
-	void finish() override;
+	/**
+	 * @brief Starts the render pass, executes the render pipeline, and then ends the render pass
+	 * @param command_buffer The command buffer to record the commands to
+	 * @param render_target The render target that is being drawn to
+	 */
+	virtual void draw_renderpass(vkb::core::HPPCommandBuffer &command_buffer, vkb::rendering::HPPRenderTarget &render_target);
+	virtual void draw_renderpass(CommandBuffer &command_buffer, RenderTarget &render_target);

```

Comment:
having two functions with different signatures like this might lead to confusion.  Why not template specialization with more conditional types?

Someone could then write 

```
template<>
VulkanSample<true>::draw_renderpass(CommandBufferType& command_buffer, RenderTargetType& render_target)`
```

That way only one signature type is visible to downstream classes for calling and overriding.

Reply:
Done.